tests/unit/invert/laplace/test_laplace_petsc3damg.cxx

```diff
@@ -0,0 +1,376 @@
+      y = i.y()/(BoutReal)ny - 0.5;
+      coef2[i] = x + y;
+    }
+    BOUT_FOR(i, mesh->getRegion3D("RGN_ALL")) {
+      x = i.x()/(BoutReal)nx - 0.5;
+      y = i.y()/(BoutReal)ny - 0.5;
+      z = i.z()/(BoutReal)nz - 0.5;
+      f3[i] = 1e3*exp(-0.5*sqrt(x*x + y*y + z*z)/sigmasq);
+      coef3[i] = x + y + sin(2*3.14159265358979323846*z);
+    }
+    auto param = GetParam();
+    forward = ForwardOperator(std::get<0>(param), std::get<1>(param),
+			      std::get<2>(param), std::get<3>(param));
+  }
+
+  ~Petsc3dAmgTest() {
+    Options::cleanup();
+    PetscErrorPrintf = PetscErrorPrintfDefault;
+  }
+
+  const BoutReal sigmasq = 0.02;
+  LaplacePetsc3dAmg solver;
+  Field2D coef2;
+  Field3D f3, coef3;
+  static constexpr BoutReal tol = 1e-8;
+  ForwardOperator forward;
+
+private:
+
+  static Options* getOptions(std::tuple<bool, bool, bool, bool> param) {
+    Options *options = Options::getRoot()->getSection("laplace");
+    (*options)["type"] = "petsc3damg";
+    (*options)["inner_boundary_flags"] = (std::get<0>(param) ? INVERT_AC_GRAD : 0) + INVERT_RHS;
+    (*options)["outer_boundary_flags"] = (std::get<1>(param) ? INVERT_AC_GRAD : 0) + INVERT_RHS;
+    (*options)["lower_boundary_flags"] = (std::get<2>(param) ? INVERT_AC_GRAD : 0) + INVERT_RHS;
+    (*options)["upper_boundary_flags"] = (std::get<3>(param) ? INVERT_AC_GRAD : 0) + INVERT_RHS;
+    (*options)["fourth_order"] = false;
+    (*options)["atol"] = tol/10; // Need to specify smaller than desired tolerance to
+    (*options)["rtol"] = tol/10; // ensure it is satisfied for every element.
+    //    (*options)["dtol"] = 1e5;
+    //    (*options)["maxits"] = 100000;
+    //    (*options)["richardson_damping_factor"] = 1.0;
+    //    (*options)["chebyshev_max"] = 100.0;
+    //    (*options)["chebyshev_min"] = 0.01;
+    //    (*options)["gmres_max_steps"] = 30;
+    return options;
+  }
+
+};
+
+INSTANTIATE_TEST_SUITE_P(LaplacePetsc3dAmgTest, Petsc3dAmgTest,
+			 testing::Values(std::make_tuple(false, false, false, false),
+					 std::make_tuple(false, false, false, true),
+					 std::make_tuple(false, false, true, false),
+					 std::make_tuple(false, true, false, false),
+					 std::make_tuple(true, false, false, false)));
+
+TEST_P(Petsc3dAmgTest, TestMatrixConstruction3D){
+  PetscMatrix<Field3D> &matrix = solver.getMatrix3D();
+  PetscVector<Field3D> vector(f3, solver.getIndexer());
+  Field3D expected = forward(f3);
+  Field3D actual = (matrix * vector).toField();
+  BOUT_FOR(i, mesh->getRegion3D("RGN_ALL")) {
+    EXPECT_NEAR(expected[i], actual[i], tol);
+  }
+}
+
+TEST_P(Petsc3dAmgTest, TestSetCoefA_2D){
+  solver.setCoefA(coef2);
+  PetscMatrix<Field3D> &matrix = solver.getMatrix3D();
+  PetscVector<Field3D> vector(f3, solver.getIndexer());
+  forward.a = coef2;
+  Field3D expected = forward(f3);
+  Field3D actual = (matrix * vector).toField();
+  BOUT_FOR(i, mesh->getRegion3D("RGN_ALL")) {
+    EXPECT_NEAR(expected[i], actual[i], tol);
+  }
```

Comment:
If you change this idiom to `EXPECT_TRUE(IsFieldEqual(expected, actual, "RGN_ALL", tol))`, the error message will also contain the index `i`, which might be useful for diagnosing the issue.

Reply:
I already took a more primitive approach to this in my docker container, just printing the coordinates of the index on each iteration. It wasn't particularly informative. There were no errors for `y == 0`, `y == 1`,  for most cases when `y == 3`, and for some cases when `y == 4`.